platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/workspace/impl/CacheBasedImmutableWorkspaceProvider.java

```diff
@@ -133,12 +104,26 @@ private static CleanupAction createCleanupAction(FileAccessTimeJournal fileAcces
     }
 
     @Override
-    public <T> T withWorkspace(String path, WorkspaceAction<T> action) {
-        return cache.withFileLock(() -> {
-            File workspace = new File(baseDirectory, path);
-            fileAccessTracker.markAccessed(workspace);
-            return action.executeInWorkspace(workspace, executionHistoryStore);
-        });
+    public ImmutableWorkspace getWorkspace(String path) {
+        File immutableWorkspace = new File(baseDirectory, path);
+        fileAccessTracker.markAccessed(immutableWorkspace);
+        return new ImmutableWorkspace() {
+            @Override
+            public File getImmutableLocation() {
+                return immutableWorkspace;
+            }
+
+            @Override
+            public <T> T withTemporaryWorkspace(TemporaryWorkspaceAction<T> action) {
+                return cache.withFileLock(() -> {
+                    // TODO Use Files.createTemporaryDirectory() instead
```

Comment:
It's among the list of followup things to do; I'll add issues for those soon.

I tried to do it as part of this PR, but it's a bigger change, because the Kotlin DSL workspace identities currently contain `/` characters, and that makes `Files.createTempDirectory()` fail. We can straighten it out, but let's do it separately.

Reply:
See https://github.com/gradle/gradle/issues/27141